MapboxNavigation/CarPlayNavigationViewController.swift

```diff
@@ -4,6 +4,96 @@ import MapboxCoreNavigation
 #if canImport(CarPlay)
 import CarPlay
 
+@available(iOS 12.0, *)
+protocol NavigationMapTemplateControllerDelegate: class {
+    func navigationMapTemplateController(_ navigationMapTemplateController: NavigationMapTemplateController, willHandle mapButton: CPMapButton)
+    func navigationMapTemplateController(_ navigationMapTemplateController: NavigationMapTemplateController, willHandle barButton: CPBarButton)
+}
+
+@available(iOS 12.0, *)
+public class NavigationMapTemplateController {
+    var mapTemplate: CPMapTemplate
+    weak var delegate: NavigationMapTemplateControllerDelegate?
+    weak var mapDelegate: CPMapTemplateDelegate? {
+        get {
+            return mapTemplate.mapDelegate
+        }
+        set {
+            mapTemplate.mapDelegate = mapDelegate
+        }
+    }
+    
+    var previousMapButtons: [CPMapButton]
+    var previousLeadingNavigationBarButtons: [CPBarButton]
+    var previousTrailingNavigationBarButtons: [CPBarButton]
+    
+    var showFeedbackButton: CPMapButton!
+    var overviewButton: CPMapButton!
+    var recenterButton: CPMapButton!
+    
+    var exitButton: CPBarButton!
+    var muteButton: CPBarButton!
+    
+    init(mapTemplate: CPMapTemplate) {
+        self.mapTemplate = mapTemplate
+        previousMapButtons = mapTemplate.mapButtons
+        previousLeadingNavigationBarButtons = mapTemplate.leadingNavigationBarButtons
+        previousTrailingNavigationBarButtons = mapTemplate.trailingNavigationBarButtons
```

Comment:
should we allow developers to mutate `self.previousTrailingNavigationBarButtons` after init?

Reply:
The idea here was to avoid unnecessarily asking the delegate to set up the buttons that were already on the navigation bar, whatever they were. (It’s possible for the user to enter route selection from at least two different screens, so there could in theory be two different sets of navigation bar buttons depending on the flow.) If we were to make these buttons mutable during the navigation session, then I guess this class would ask CarPlayNavigationViewController to ask its delegate for the buttons.

Reply:
Going to merge this PR as a first step and streamline button management directly on the carplay branch as tail work.